File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValid(self, root: TreeNode, left: int, right: int) -> bool:
        if not root:
            return True
        if not (left < root.val < right):
            return False
        return self.isValid(root.left, left, root.val) and self.isValid(root.right, root.val, right)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        return self.isValid(root, float("-inf"), float("+inf"))        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {

        // BFS - In Order
        if (root == null) {
            return -1;
        }
        int ans = 0;

        ArrayDeque<TreeNode> stack = new ArrayDeque<>();
        
        while (!stack.isEmpty() || root != null) {
            while(root != null) {
                stack.push(root);
                root = root.left;
            }
            TreeNode node = stack.pop();
            k = k - 1;
            if (k == 0){
                ans = node.val;
                break;
            }
            if(node.right != null) {
                stack.push(node.right);
            }
        }
        return ans;
    }
}
